File: android/app/src/main/java/com/pearsonmedia/loci/util/HapticManager.kt
package com.pearsonmedia.loci.util

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages haptic feedback for user interactions.
 * Matches iOS HapticManager pattern.
 */
@Singleton
class HapticManager @Inject constructor(
    private val context: Context
) {
    private val vibrator: Vibrator by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            manager.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }
    }

    fun light() {
        vibrate(VibrationEffect.createOneShot(30, VibrationEffect.DEFAULT_AMPLITUDE))
    }

    fun medium() {
        vibrate(VibrationEffect.createOneShot(50, VibrationEffect.DEFAULT_AMPLITUDE))
    }

    fun heavy() {
        vibrate(VibrationEffect.createOneShot(80, VibrationEffect.DEFAULT_AMPLITUDE))
    }

    fun success() {
        val timings = longArrayOf(0, 50, 50, 100)
        val amplitudes = intArrayOf(0, 80, 0, 160)
        vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
    }

    fun error() {
        val timings = longArrayOf(0, 100, 50, 100, 50, 100)
        val amplitudes = intArrayOf(0, 200, 0, 200, 0, 200)
        vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
    }

    fun selection() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            vibrate(VibrationEffect.createPredefined(VibrationEffect.EFFECT_CLICK))
        } else {
            light()
        }
    }

    // --- Premium sequences (US-169) ---

    /**
     * Two-beat "lock-in" feel when a recording starts: medium impact
     * followed by a light tap ~90ms later.
     */
    fun recordStartSequence() {
        val timings = longArrayOf(0, 50, 40, 20)
        val amplitudes = intArrayOf(0, 180, 0, 90)
        vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
    }

    /**
     * Light tap priming the user before a success chime — used when a
     * locus is saved.
     */
    fun saveSuccessSequence() {
        val timings = longArrayOf(0, 25, 90, 40, 40, 100)
        val amplitudes = intArrayOf(0, 90, 0, 140, 0, 200)
        vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
    }

    /**
     * Selection → heavy impact → success celebration. Used when a paid
     * tier is unlocked or a household invite is accepted.
     */
    fun unlockSequence() {
        val timings = longArrayOf(0, 20, 80, 80, 120, 60, 60, 100)
        val amplitudes = intArrayOf(0, 80, 0, 220, 0, 140, 0, 200)
        vibrate(VibrationEffect.createWaveform(timings, amplitudes, -1))
    }

    private fun vibrate(effect: VibrationEffect) {
        if (vibrator.hasVibrator()) {
            vibrator.vibrate(effect)
        }
    }
}

File: Lociate/Utilities/HapticManager.swift
import UIKit

enum HapticManager {
    // MARK: - Recording

    static func recordStart() {
        let generator = UIImpactFeedbackGenerator(style: .medium)
        generator.prepare()
        generator.impactOccurred()
    }

    static func recordStop() {
        let generator = UIImpactFeedbackGenerator(style: .light)
        generator.prepare()
        generator.impactOccurred()
    }

    // MARK: - Actions

    static func saveSuccess() {
        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(.success)
    }

    static func archive() {
        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(.warning)
    }

    static func delete() {
        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(.error)
    }

    // MARK: - Navigation

    static func tabSelection() {
        let generator = UISelectionFeedbackGenerator()
        generator.prepare()
        generator.selectionChanged()
    }

    // MARK: - Premium Sequences (US-169)

    /// Two-beat confirmation when a recording starts: medium impact followed
    /// by a light impact 90ms later. Creates a "lock-in" feel that makes the
    /// start of a recording feel intentional and premium.
    static func recordStartSequence() {
        let primary = UIImpactFeedbackGenerator(style: .medium)
        primary.prepare()
        primary.impactOccurred()
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.09) {
            let secondary = UIImpactFeedbackGenerator(style: .light)
            secondary.prepare()
            secondary.impactOccurred()
        }
    }

    /// Light impact followed by a success notification — used when a locus
    /// is saved. The light tap primes the user before the success chime.
    static func saveSuccessSequence() {
        let tap = UIImpactFeedbackGenerator(style: .light)
        tap.prepare()
        tap.impactOccurred()
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.12) {
            let success = UINotificationFeedbackGenerator()
            success.prepare()
            success.notificationOccurred(.success)
        }
    }

    /// Selection → heavy impact → success — used when a paid tier is
    /// unlocked or a household invite is accepted. Feels celebratory.
    static func unlockSequence() {
        let selection = UISelectionFeedbackGenerator()
        selection.prepare()
        selection.selectionChanged()
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            let impact = UIImpactFeedbackGenerator(style: .heavy)
            impact.prepare()
            impact.impactOccurred()
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.24) {
            let success = UINotificationFeedbackGenerator()
            success.prepare()
            success.notificationOccurred(.success)
        }
    }
}
